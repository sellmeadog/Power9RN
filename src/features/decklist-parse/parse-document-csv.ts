import Papa from 'papaparse';

import { P9CreateDecklistEntryInfo, P9DocumentParseResult } from './parse-document-types';

const KNOWN_COUNT_HEADERS = ['quantity', 'amount', 'count', 'total qty', 'qty', 'regular'];
const KNOWN_NAME_HEADERS = ['card', 'card name', 'name', 'card_name'];
const KNOWN_EXPANSION_CODE_HEADERS = ['set', 'edition', 'expansion symbol', 'set id', 'set code', 'set_code', 'code'];
const KNOWN_COLLECTOR_NUMBER_HEADERS = ['collector #', 'card number'];

/**
 * Alias for each row in the parsed CSV file
 */
type P9CSVDataRow = Partial<P9CreateDecklistEntryInfo>;

/**
 * Adds additional typing to the default Papa.ParseResult interface
 */
interface P9CSVParseResult extends Papa.ParseResult<P9CSVDataRow> {}

/**
 * Initiates the parsing of a CSV document content string and returns a P9DocumentParseResult when complete
 * @param csvString The document content read from a parent CSV file
 */
export const parseDocumentCsv = (csvString: string): P9DocumentParseResult => {
  const parseResult = parseCSVString(csvString);
  const [manualEntries, parsedEntries] = reduceDataRows(parseResult);

  return {
    manualEntries: manualEntries.join('\n'),
    parsedEntries,
  };
};

/**
 * Reduces a single P9CSVDataRow into a string and P9CreateDecklistEntryInfo representation if possible
 * @param param0 The destructured aggregate value from the parent reduce function
 * @param param1 The next P9CSVDataRow to reduce
 */
const reduceDataRow = (
  [manual, parsed]: [string[], P9CreateDecklistEntryInfo[]],
  { count, cardName, expansionCode, collectorNumber }: P9CSVDataRow,
): [string[], P9CreateDecklistEntryInfo[]] => {
  if (!count || !cardName || /^\W*?$/gim.test(cardName)) {
    return [manual, parsed];
  }

  count = count.trim();
  cardName = cardName.trim();

  const builder = [count, cardName];
  const entry: P9CreateDecklistEntryInfo = { type: 'maindeck', count, cardName };

  if (expansionCode) {
    expansionCode = expansionCode.trim();
    builder.push(`(${expansionCode})`);
    entry.expansionCode = expansionCode;
  }

  if (collectorNumber) {
    if (collectorNumber.indexOf('/') > -1) {
      collectorNumber = collectorNumber.slice(0, collectorNumber.indexOf('/'));
    }
    collectorNumber = collectorNumber.trim();
    builder.push(collectorNumber);
    entry.collectorNumber = collectorNumber;
  }

  return [manual.concat([builder.join(' ')]), parsed.concat([entry])];
};

/**
 * Reduces the raw Papa.ParseResult
 * @param result The enhanced Papa.ParseResult to be reduced
 */
const reduceDataRows = (result: P9CSVParseResult): [string[], P9CreateDecklistEntryInfo[]] => {
  const manualEntries: string[] = [];
  const parsedEntries: P9CreateDecklistEntryInfo[] = [];

  return result.data.reduce(reduceDataRow, [manualEntries, parsedEntries] as [string[], P9CreateDecklistEntryInfo[]]);
};

/**
 * Parses the given CSV data string into an array of P9CSVDataRow objects
 * @param csvString The CSV string to parse
 */
const parseCSVString = (csvString: string): P9CSVParseResult => {
  return Papa.parse(csvString, {
    header: true,
    skipEmptyLines: true,
    transformHeader: (header) => {
      header = header.toLowerCase();
      if (KNOWN_COUNT_HEADERS.includes(header)) {
        return 'count';
      }
      if (KNOWN_NAME_HEADERS.includes(header)) {
        return 'cardName';
      }
      if (KNOWN_EXPANSION_CODE_HEADERS.includes(header)) {
        return 'expansionCode';
      }
      if (KNOWN_COLLECTOR_NUMBER_HEADERS.includes(header)) {
        return 'collectorNumber';
      }
      return `UNUSED_${header}`;
    },
  });
};
