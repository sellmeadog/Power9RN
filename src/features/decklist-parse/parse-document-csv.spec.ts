import { parseDocumentCsv } from './parse-document-csv';

describe('parseDocumentCsv', () => {
  it('should parse MTGO', () => {
    // Arrange
    const csvString = `Card Name,Quantity,ID #,Rarity,Set,Collector #,Premium
"Banisher Priest",1,51909,Uncommon,PRM,1136/1158,Yes
"Batterskull",10,51909,Uncommon,PRM,1136/1158,Yes
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse MTG Studio', () => {
    // Arrange
    const csvString = `Name,Edition,Qty,Foil
Aether Vial,MMA,1,No
Anax and Cymede,THS,4,Yes
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse Deckbox', () => {
    // Arrange
    const csvString = `Count,Tradelist Count,Name,Edition,Card Number,Condition,Language,Foil,Signed,Artist Proof,Altered Art,Mis,
4,4,Angel of Serenity,Return to Ravnica,1,Near Mint,English,,,,,,,,
1,1,Ashen Rider,Theros,187,Near Mint,English,,,,,,,,
1,0,Anax and Cymede,Theros,186,Near Mint,English,foil,,,,,,,
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse Decked Builder', () => {
    // Arrange
    const csvString = `Total Qty,Reg Qty,Foil Qty,Card,Set,Mana Cost,Card Type,Color,Rarity,Mvid,Single Price,Single Foil Price,Total Price,Price Source,Notes
3,2,1,Black Sun's Zenith,Mirrodin Besieged,XBB,Sorcery,Black,Rare,214061,1.00,6.25,7.25,cardshark,
1,1,0,Snapcaster Mage,Innistrad,1U,Creature  - Human Wizard,Blue,Rare,227676,26.60,115.00,141.60,cardshark,
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse Pucatrade', () => {
    // Arrange
    const csvString = `Count,Name,Edition,Rarity,Expansion Symbol,Points,Foil,Condition,Language,Status,Entered Date,Updated Date,Exported Date
1,Breeding Pool,Gatecrash,RARE,GTC,1009,0,"Near Mint","English",NOT FOR TRADE,"11/19/2014","11/29/2014","11/30/2014",19632
8,Rattleclaw Mystic,Khans of Tarkir,RARE,KTK,194,0,"Near Mint","English",HAVE,"11/29/2014","11/29/2014","11/30/2014",25942
1,Opulent Palace,Khans of Tarkir,UNCOMMON,KTK,461,1,"Near Mint","English",HAVE,"11/29/2014","11/29/2014","11/30/2014",25928
1,Tasigur the Golden Fang,Fate Reforged,RARE,FRF,1045,1,"Near Mint","English",HAVE,"1/24/2015","1/24/2015","2/03/2015",270
1,Tasigur the Golden Fang,Fate Reforged,RARE,FRF,1045,1,"Near Mint","English",HAVE,"1/24/2015","1/24/2015","2/03/2015",270
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse MTGGoldfish', () => {
    // Arrange
    const csvString = `Card,Set ID,Set Name,Quantity,Foil
Aether Vial,MMA,Modern Masters,1
Anax and Cymede,THS,Theros,4,FOIL
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse MTGStocks', () => {
    // Arrange
    const csvString = `"Card","Set","Quantity","Price","Condition","Language","Foil","Signed"
"Abandon Hope","Tempest",2,0.24,M,en,No,No
"Abduction","Classic Sixth Edition",1,0.33,M,en,No,No
"Advent of the Wurm","Modern Masters 2017",1,0.99,M,en,Yes,No
"Advent of the Wurm","Modern Masters 2017",3,0.35,M,en,No,No
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse TCGplayer', () => {
    // Arrange
    const csvString = `Quantity,Name,Simple Name,Set,Card Number,Set Code,Printing,Condition,Language,Rarity,Product ID,SKU
1,Verdant Catacombs,Verdant Catacombs,Zendikar,229,ZEN,Normal,Near Mint,English,Rare,33470,315319
1,Graven Cairns,Graven Cairns,Zendikar Expeditions,28,EXP,Foil,Near Mint,English,Mythic,110729,3042202
1,Olivia Voldaren,Olivia Voldaren,Innistrad,215,ISD,Foil,Near Mint,English,Mythic,52181,500457
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse Deckstats.net', () => {
    // Arrange
    const csvString = `amount,card_name,is_foil,is_pinned,set_id,set_code
1,"Abandon Reason",0,0,147,"EMN"
2,"Abandoned Sarcophagus",0,0,187,"HOU"
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse MTG Manager', () => {
    // Arrange
    const csvString = `Quantity,Name,Code,PurchasePrice,Foil,Condition,Language,PurchaseDate
1,"Amulet of Vigor",WWK,18.04,0,0,0,5/6/2018
1,"Arcane Lighthouse",C14,3.83,0,0,0,5/6/2018
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse count from Regular and Foil headers', () => {
    // Arrange
    const csvString = `Set,Card,Regular,Foil
Amonkhet,Forest,3,1
`;

    // Act
    const actual = parseDocumentCsv(csvString);

    // Assert
    expect(actual).toMatchSnapshot();
  });
});
