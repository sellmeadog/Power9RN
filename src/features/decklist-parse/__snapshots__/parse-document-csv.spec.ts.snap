// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseDocumentCsv should parse Deckbox 1`] = `
Object {
  "manualEntries": "4 Angel of Serenity (Return to Ravnica) 1
1 Ashen Rider (Theros) 187
1 Anax and Cymede (Theros) 186",
  "parsedEntries": Array [
    Object {
      "cardName": "Angel of Serenity",
      "collectorNumber": "1",
      "count": "4",
      "expansionCode": "Return to Ravnica",
      "type": "maindeck",
    },
    Object {
      "cardName": "Ashen Rider",
      "collectorNumber": "187",
      "count": "1",
      "expansionCode": "Theros",
      "type": "maindeck",
    },
    Object {
      "cardName": "Anax and Cymede",
      "collectorNumber": "186",
      "count": "1",
      "expansionCode": "Theros",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse Decked Builder 1`] = `
Object {
  "manualEntries": "3 Black Sun's Zenith (Mirrodin Besieged)
1 Snapcaster Mage (Innistrad)",
  "parsedEntries": Array [
    Object {
      "cardName": "Black Sun's Zenith",
      "count": "3",
      "expansionCode": "Mirrodin Besieged",
      "type": "maindeck",
    },
    Object {
      "cardName": "Snapcaster Mage",
      "count": "1",
      "expansionCode": "Innistrad",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse Deckstats.net 1`] = `
Object {
  "manualEntries": "1 Abandon Reason (EMN)
2 Abandoned Sarcophagus (HOU)",
  "parsedEntries": Array [
    Object {
      "cardName": "Abandon Reason",
      "count": "1",
      "expansionCode": "EMN",
      "type": "maindeck",
    },
    Object {
      "cardName": "Abandoned Sarcophagus",
      "count": "2",
      "expansionCode": "HOU",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse MTG Manager 1`] = `
Object {
  "manualEntries": "1 Amulet of Vigor (WWK)
1 Arcane Lighthouse (C14)",
  "parsedEntries": Array [
    Object {
      "cardName": "Amulet of Vigor",
      "count": "1",
      "expansionCode": "WWK",
      "type": "maindeck",
    },
    Object {
      "cardName": "Arcane Lighthouse",
      "count": "1",
      "expansionCode": "C14",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse MTG Studio 1`] = `
Object {
  "manualEntries": "1 Aether Vial (MMA)
4 Anax and Cymede (THS)",
  "parsedEntries": Array [
    Object {
      "cardName": "Aether Vial",
      "count": "1",
      "expansionCode": "MMA",
      "type": "maindeck",
    },
    Object {
      "cardName": "Anax and Cymede",
      "count": "4",
      "expansionCode": "THS",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse MTGGoldfish 1`] = `
Object {
  "manualEntries": "1 Aether Vial (MMA)
4 Anax and Cymede (THS)",
  "parsedEntries": Array [
    Object {
      "cardName": "Aether Vial",
      "count": "1",
      "expansionCode": "MMA",
      "type": "maindeck",
    },
    Object {
      "cardName": "Anax and Cymede",
      "count": "4",
      "expansionCode": "THS",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse MTGO 1`] = `
Object {
  "manualEntries": "1 Banisher Priest (PRM) 1136
10 Batterskull (PRM) 1136",
  "parsedEntries": Array [
    Object {
      "cardName": "Banisher Priest",
      "collectorNumber": "1136",
      "count": "1",
      "expansionCode": "PRM",
      "type": "maindeck",
    },
    Object {
      "cardName": "Batterskull",
      "collectorNumber": "1136",
      "count": "10",
      "expansionCode": "PRM",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse MTGStocks 1`] = `
Object {
  "manualEntries": "2 Abandon Hope (Tempest)
1 Abduction (Classic Sixth Edition)
1 Advent of the Wurm (Modern Masters 2017)
3 Advent of the Wurm (Modern Masters 2017)",
  "parsedEntries": Array [
    Object {
      "cardName": "Abandon Hope",
      "count": "2",
      "expansionCode": "Tempest",
      "type": "maindeck",
    },
    Object {
      "cardName": "Abduction",
      "count": "1",
      "expansionCode": "Classic Sixth Edition",
      "type": "maindeck",
    },
    Object {
      "cardName": "Advent of the Wurm",
      "count": "1",
      "expansionCode": "Modern Masters 2017",
      "type": "maindeck",
    },
    Object {
      "cardName": "Advent of the Wurm",
      "count": "3",
      "expansionCode": "Modern Masters 2017",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse Pucatrade 1`] = `
Object {
  "manualEntries": "1 Breeding Pool (GTC)
8 Rattleclaw Mystic (KTK)
1 Opulent Palace (KTK)
1 Tasigur the Golden Fang (FRF)
1 Tasigur the Golden Fang (FRF)",
  "parsedEntries": Array [
    Object {
      "cardName": "Breeding Pool",
      "count": "1",
      "expansionCode": "GTC",
      "type": "maindeck",
    },
    Object {
      "cardName": "Rattleclaw Mystic",
      "count": "8",
      "expansionCode": "KTK",
      "type": "maindeck",
    },
    Object {
      "cardName": "Opulent Palace",
      "count": "1",
      "expansionCode": "KTK",
      "type": "maindeck",
    },
    Object {
      "cardName": "Tasigur the Golden Fang",
      "count": "1",
      "expansionCode": "FRF",
      "type": "maindeck",
    },
    Object {
      "cardName": "Tasigur the Golden Fang",
      "count": "1",
      "expansionCode": "FRF",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse TCGplayer 1`] = `
Object {
  "manualEntries": "1 Verdant Catacombs (ZEN) 229
1 Graven Cairns (EXP) 28
1 Olivia Voldaren (ISD) 215",
  "parsedEntries": Array [
    Object {
      "cardName": "Verdant Catacombs",
      "collectorNumber": "229",
      "count": "1",
      "expansionCode": "ZEN",
      "type": "maindeck",
    },
    Object {
      "cardName": "Graven Cairns",
      "collectorNumber": "28",
      "count": "1",
      "expansionCode": "EXP",
      "type": "maindeck",
    },
    Object {
      "cardName": "Olivia Voldaren",
      "collectorNumber": "215",
      "count": "1",
      "expansionCode": "ISD",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocumentCsv should parse count from Regular and Foil headers 1`] = `
Object {
  "manualEntries": "3 Forest (Amonkhet)",
  "parsedEntries": Array [
    Object {
      "cardName": "Forest",
      "count": "3",
      "expansionCode": "Amonkhet",
      "type": "maindeck",
    },
  ],
}
`;
